File: solutions/php/killer-sudoku-helper/1/KillerSudokuHelper.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

class KillerSudokuHelper
{
    public function combinations(int $sum, int $size, array $exclude): array
    {
        $result = [];
        // Get available digits (1-9) excluding the forbidden ones
        $available = array_values(array_diff(range(1, 9), $exclude));
        
        // Use backtracking to find all valid combinations
        $this->backtrack($available, $size, $sum, 0, [], $result);
        
        return $result;
    }
    
    private function backtrack(array $available, int $size, int $target, int $start, array $current, array &$result): void
    {
        // Base case: if we have selected the required number of digits
        if (count($current) === $size) {
            // If the sum matches, we found a valid combination
            if ($target === 0) {
                $result[] = $current;
            }
            return;
        }
        
        // Try each available digit starting from $start index
        for ($i = $start; $i < count($available); $i++) {
            $digit = $available[$i];
            
            // Pruning: if digit is greater than remaining target, 
            // all subsequent digits will also be too large (since array is sorted)
            if ($digit > $target) {
                break;
            }
            
            // Include this digit in the current combination
            $current[] = $digit;
            
            // Recurse with reduced target and next starting index
            // (i+1 ensures we don't reuse the same digit)
            $this->backtrack($available, $size, $target - $digit, $i + 1, $current, $result);
            
            // Backtrack: remove the last added digit to try other possibilities
            array_pop($current);
        }
    }
}

File: solutions/java/killer-sudoku-helper/1/src/main/java/KillerSudokuHelper.java
import java.util.ArrayList;
import java.util.List;

public class KillerSudokuHelper {

    // Main method: find combinations with exclusions
    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize, List<Integer> exclude) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        
        // Start backtracking from digit 1
        backtrack(result, current, cageSum, cageSize, exclude, 1);
        
        return result;
    }

    // Overload: find combinations without exclusions
    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize) {
        return combinationsInCage(cageSum, cageSize, new ArrayList<>());
    }
    
    // Backtracking helper to generate valid combinations
    private void backtrack(List<List<Integer>> result, List<Integer> current, 
                          int remainingSum, int remainingSize, List<Integer> exclude, int start) {
        // Base case: found a valid combination
        if (remainingSize == 0) {
            if (remainingSum == 0) {
                result.add(new ArrayList<>(current));
            }
            return;
        }
        
        // Try each digit from start to 9
        for (int digit = start; digit <= 9; digit++) {
            // Skip excluded digits
            if (exclude.contains(digit)) {
                continue;
            }
            
            // Prune: digit too large for remaining sum
            if (digit > remainingSum) {
                continue;
            }
            
            // Choose: add digit to current combination
            current.add(digit);
            
            // Explore: recurse with next digit (digit+1 ensures no duplicates)
            backtrack(result, current, remainingSum - digit, remainingSize - 1, exclude, digit + 1);
            
            // Unchoose: backtrack for next iteration
            current.remove(current.size() - 1);
        }
    }
}